[JENKINS-5597] add option to follow symlinks during artifact archiving

mwinter69 opened the pull request:
add an option to ArtifactArchiver that allows to configure follow symlink behaviour of the DirectoryScanner used by ArtifactArchiver. For newly created steps the default is now to not follow symlinks. Existing steps will work as before (also for pipeline).

See [JENKINS-5597](https://issues.jenkins-ci.org/browse/JENKINS-5597).

Minor improvement.

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Entry 1: Add option to configure follow symlinks when archiving artifacts

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

mwinter69 committed:
[JENKINS-5597] don't follow symlinks

add an option that allows to configure the DirectoryScanner used by
ArtifactArchiver to not follow symbolic links. Option is enabled by
default for backwards compatibility.

mwinter69 committed:
missing readresolve

need to resolve in case it is not defined

Wadeck reviewed (commented):
I am not sure we really want to have such feature, especially for security reasons. In the recent patch around illegal symlink (trying to escape the workspace) there was a discussion around removing completely the symlink feature to avoid such troubles.

@jglick any thoughts here?

mwinter69 commented:
actually it is now possible to disable the follow symlink. Before it was always enabled, causing some issues.

daniel-beck commented:
If there is a security issue here, then it already exists today: _Invalid_ symlinks are archived as symlinks currently, so this should be no different from declaring all symlinks to be invalid today. (In fact, that was [SECURITY-162](https://jenkins.io/security/advisory/2015-02-27/#security-162-directory-traversal-from-artifacts-via-symlink)).

LakshanSS commented on core/src/main/resources/hudson/tasks/ArtifactArchiver/config.jelly:
```diff
@@ -46,5 +46,8 @@ THE SOFTWARE.
+      <f:checkbox title="${%followSymlinks}" default="true"/>
+    </f:entry>
   </f:advanced>
 </j:jelly>
```
Have to add a newline in the end of file

LakshanSS commented on core/src/main/resources/hudson/tasks/ArtifactArchiver/config.properties:
```diff
@@ -23,4 +23,5 @@
-caseSensitive=Treat include and exclude patterns as case sensitive
\ No newline at end of file
+caseSensitive=Treat include and exclude patterns as case sensitive
+followSymlinks=Follow symbolic links
```
Can add a new line in the end of file

LakshanSS commented on core/src/main/resources/hudson/tasks/ArtifactArchiver/config_de.properties:
```diff
@@ -27,3 +27,4 @@ allowEmptyArchive=Den Build nicht als fehlgeschlagen markieren, wenn nichts arch
 caseSensitive=Gro\u00DF- und Kleinschreibung im Dateisuchmuster ber\u00FCcksichtigen
 defaultExcludes=Standard-Ausschlussmuster verwenden
 onlyIfSuccessful=Artefakte nur archivieren, wenn der Build erfolgreich ist
+followSymlinks=Symbolischen Links folgen
```
Add a new line in the end of the file

LakshanSS commented on core/src/main/resources/hudson/tasks/ArtifactArchiver/help-followSymlinks.html:
```diff
@@ -0,0 +1,5 @@
+    Artifact archiver uses Ant <code>org.apache.tools.ant.DirectoryScanner</code> which by default follows symbolic links.
+    For instance, if the job produces a file a.log and a symbolic link b.log -> a.log, the pattern "*.log" will archive both files.<BR/><BR/>
+    This option can be used to disable following of symbolic links. When it's unchecked, pattern "*.log" will match only a.log but no b.log.
+</div>
```
add a new line in the end of the file

LakshanSS commented on core/src/main/resources/hudson/tasks/ArtifactArchiver/help-followSymlinks_de.html:
```diff
@@ -0,0 +1,5 @@
+    Der Archivierer benutzt von Ant die Klasse <code>org.apache.tools.ant.DirectoryScanner</code> welche standardmäßig symbolische Links nachverfolgt.
+    Wenn als Artefakte z.B. die Datei a.log sowie ein symbolischer Link b.log -> a.log erzeugt werden, dann wird das Muster "*.log" beides archivieren.<BR/><BR/>
+    Diese Option kann benutzt werden um das Nachverfolgen von symbolischen Links abzuschalten. Falls nicht gesetzt, dann wird das Muster "*.log" nur a.log archivieren aber nicht b.log. 
+</div>
```
add a new line in the end of the file

jglick commented:
Indeed going forward I recommend skipping all symlinks (either reporting an error or logging a warning). Any project which actually needs to record a directory tree containing symlinks or other filesystem exotica is better advised to create a tarball in the workspace as part of the build script.

AbhishekSharma20 commented:
Hi @jglick @Wadeck ,
Can you please integrate these changes? It would be very helpful. Our work area uses recursive symlinks as a feature and cannot be removed since there are a lot of dependencies. There're many people affected by this issue, we have to put the manual effort in building custom jenkins.

If there is an enhancement at a later stage regarding removing symlinks traversal all together, then just changing the default value of the boolean variable controlling this behavior will serve the purpose.

-Abhishek

jglick commented:
> we have to put the manual effort in building custom jenkins.

Seems unlikely; again, you can already just create a tarball during your build and archive that instead.

To make this more attractive and guide users to do the right thing, perhaps we could just offer an option to `ArtifactArchiver`. Something like:

```groovy
sh 'make build'
dir('build') {
  archiveArtifacts includes: '**/', archive: 'binaries.tgz'
}
```

which would do the tarring on the agent side, saving you the trouble of adding a `tar` or `zip` command to your build script. (The compression method could be inferred by the filename: `.zip`, `.tar`, `.tar.gz` / `.tgz`.) We could then consider making it an error if a symlink is encountered _without_ this option being set, with a system property to keep the problematic original behavior for a while before really deleting that ability.

AbhishekSharma20 commented:
Hi @jglick,
My issue was that my directory is structured in a very complex way containing many levels of symlinks (because of business requirements). archiving this way won't work as the workspace tree will still be created inside Jenkins. and it is that object (the Hashmap containing directoryStructure) which is getting infinite bloating up the memory. As of now, we have faked the work area to Jenkins to an empty directory inside our actual work area. Since this directory will not contain anything but artifacts to be archived, this issue will not come.

MRamonLeon reviewed (approved):
I'm fine with this change, the discussion about the use or not of symlinks is another topic I believe. So approving.

It would be great if you add the EOL suggested.

fcojfernandez commented:
After reading @jglick , @Wadeck and @daniel-beck 's comment I have actual doubts if this PR should be merged or not. Maybe it's worth to open a discussion in the mailing list?

alecharp commented:
I personally don't feel comfortable with this feature, I can only see security risks. I would prefer not to have this in Jenkins, symlinks in general don't feel like the correct solution for a build.

mwinter69 commented:
We have one case where a node build is creating many softlinks to other modules (even with cycles). Using find to get the the real number of files and folder returns like 63000 lines.
But due to the softlinking this is blown up to almost 1.3 million. In combination with wildcard patterns we see OOM exceptions in the ant directory scanner as also reported by @AbhishekSharma20 . With this feature there is no issue.

I agree that symlinks are dangerous when they lead to files outside of the workspace. It should not be possible to archive anything that is outside of the workspace (though it will not stop anyone to just copy the file instead of creating a softlink and archive the file this way).
But symlinks that stay within the workspace are totally ok I think and should not be forbidden. 

The solution by @jglick with an automatic archiving will not change anything to the OOM if the directory scanner is enabled to follow symlinks.

My PR will not make it any worse to the current situation but solves a problem when there are symlinks in the workspace.

mwinter69 committed:
add linebreaks at end of file

Wadeck commented:
I will try to spend some time to re-review this PR as it seems promising :)

jglick commented:
> the workspace tree will still be created inside Jenkins. and it is that object (the Hashmap containing directoryStructure) which is getting infinite bloating up the memory

This just sounds like a bug. Are there steps to reproduce?

> As of now, we have faked the work area to Jenkins to an empty directory inside our actual work area. Since this directory will not contain anything but artifacts to be archived, this issue will not come.

That seems like a good workaround.

jeffret-b commented:
I plan on reviewing in the next couple of days.

varyvol commented on core/src/main/java/hudson/tasks/ArtifactArchiver.java:
```diff
@@ -110,6 +110,12 @@
     private Boolean caseSensitive = true;
 
+    /**
+     * Indicate whether include and exclude patterns should be considered as case sensitive
```
```suggestion
     * Indicate whether symbolic links should be followed or not
```

varyvol reviewed (approved):
Minor suggestion for the Javadoc, but the change looks good give it actually allows to disable following symlinks, which could make the situation better.

jeffret-b commented:
I don't see that this PR causes any additional problems that don't already exist. It provides some additional capability which can be useful. I'm in favor of accepting this PR (with a little bit of change).

Ultimately, I agree with Jesse and others that handling symlinks is just a bad idea. Tar is a much better solution for archiving symlinks. Zip doesn't really know what to do with symlinks, especially on Windows.

But, a lot of people use the Archive Artifact capability already and this would be a step in the right direction to better use. I would prefer to make not following symlinks the default, but that is probably too much of a change in behavior. Maybe we could change the default later on.

I don't see that we need to hold up this PR for some better, more comprehensive solution that doesn't yet exist.

jeffret-b reviewed (changes requested):
This behavior really deserves an automated test. It looks like there are existing tests in ArtifactArchiverTest that should be easily modifiable to test this additional behavior. I can help suggest what to do if needed.

I favor approving this PR and merging it in, but I would like to have a test for it.

jeffret-b commented on core/src/main/java/hudson/tasks/ArtifactArchiver.java:
```diff
@@ -110,6 +110,12 @@
+     * Indicate whether include and exclude patterns should be considered as case sensitive
+     */
+    @Nonnull
+    private Boolean followSymlinks = true;
```
@daniel-beck, if this is false and with the readResolve() setting to true on null, does that retain old behavior for existing projects and default to not following symlinks for new ones?

We should definitely have that behavior and then I'm even more in strongly in favor of this PR.

daniel-beck replied:
@jeffret-b Mostly.

This being `Boolean` and `readResolve` combine to set the the default when loading existing projects. This is definitely the correct approach to not break everyone's existing projects.

This sets the programmatic default for new instances, but the UI needs to be considered as well. https://github.com/jenkinsci/jenkins/pull/3947/files#diff-16a525c7f3ec8f3f359bc969708f920cR50 would make a different value here next to useless (modulo Pipeline snippet generator and such).

jeffret-b replied:
So, change that line to `false` also, and it should work, right? That's good.

jeffret-b replied:
If I understand the flow of discussion correctly, to maintain backwards compatibility for pipeline, we need to keep this as `true` and keep the new UI setting in config.jelly defaulting to false.

daniel-beck commented on core/src/main/java/hudson/tasks/ArtifactArchiver.java:
```diff
@@ -153,6 +159,9 @@ protected Object readResolve() {
         }
+        if (followSymlinks == null) {
+          followSymlinks = true;
+      }
```
```suggestion
        }
```

daniel-beck commented on core/src/main/java/hudson/tasks/ArtifactArchiver.java:
```diff
@@ -214,6 +223,14 @@ public boolean isCaseSensitive() {
     }
 
+    public boolean isFollowSymlinks() {
+        return followSymlinks;
```
How confident are we that this never breaks thanks to `readResolve`?

jeffret-b replied:
I'm pretty confident. If we had another test or two, I'd be sufficiently confident.

daniel-beck commented on core/src/main/java/hudson/tasks/ArtifactArchiver.java:
```diff
@@ -214,6 +223,14 @@ public boolean isCaseSensitive() {
+    }
+
+    @DataBoundSetter public final void setFollowSymlinks(boolean followSymlinks) {
+      this.followSymlinks = followSymlinks;
```
```suggestion
        this.followSymlinks = followSymlinks;
```

daniel-beck commented:
Overall I like this a lot. I would even argue (as does @jeffret-b ) that not following symlinks is a superior approach to doing so.

Given how central archiving of artifacts is to Jenkins, this deserves autotests. Otherwise I see nothing wrong with this.

mwinter69 committed:
add test for not follow symlinks

mwinter69 committed:
Merge branch 'master' into JENKINS-5597

mwinter69 committed:
fix indentation

mwinter69 committed:
add tests

update ant to 1.10.7 which supports symlinks on windows and is able to
properly ignore them

mwinter69 commented:
added some tests.
I included and update of ant to 1.10.7 as since 1.10.2 the ant directory scanner is able to properly detect symlinks on windows.
Haven't yet tested behaviour of pipeline jobs

mwinter69 committed:
change default in jelly to false

mwinter69 committed:
fix test execution

jeffret-b commented:
I'm glad to see the addition of tests. I plan on reviewing this again tomorrow.

daniel-beck commented:
@jglick Is there guidance on how Pipeline steps should evolve to add options with different default behavior?

For traditional job types, we can `readResolve` into legacy behavior, but if we change the new field `followSymlinks` to `false`, IIUC, the `archiveArtifacts` step will change behavior.

Having a different default behavior for new freestyle jobs vs. new pipelines (all pipelines really) is also weird.

jglick commented:
> Is there guidance on how Pipeline steps should evolve to add options with different default behavior?

The [usual guide](https://jenkins.io/doc/developer/plugin-development/pipeline-integration/#handling-default-values).

I have not followed the discussion here about default values. Whatever the constructor picks, that will be the default value for Pipeline. I would certainly not advise changing the default behavior, as you could break existing scripts. I have no opinion about the behavior for non-Pipeline projects.

jeffret-b commented on core/src/main/resources/hudson/tasks/ArtifactArchiver/help-followSymlinks.html:
```diff
@@ -0,0 +1,6 @@
+<div>
```
I recommend simplifying this. I don't think the user really needs to know about Ant and Directory Scanner. I would probably remove the help text entirely, as I think "Follow symbolic links" is sufficiently explanatory.

jeffret-b reviewed (approved):
I'm glad this now has the two new tests. That helps a lot.

As I'm understanding it, this now works correctly to maintain compatibility with existing jobs and with pipelines but defaults in the UI to not follow symlinks. That's probably as good as we can get for reducing symlink usage at this point.

I left a comment suggesting simplifying or removing the help text.

I support this change. It will move us towards reducing symlink usage, which is a good thing.

daniel-beck commented:
> keep the new UI setting in config.jelly defaulting to false.

That's the question – it looks like we cannot just change the programmatic default (thanks pipeline!); I wonder what's less bad:

* Having the inferior behavior be the default for all job types (terrible but consistent)
* Having the inferior behavior only be the default for pipelines (terribly inconsistent)

The former is probably better. Even if it still sucks.

jeffret-b commented:
>     * Having the inferior behavior be the default for all job types (terrible but consistent)
> 
>     * Having the inferior behavior only be the default for pipelines (terribly inconsistent)
> 
> 
> The former is probably better. Even if it still sucks.

+1 for the second choice

jglick commented:
> it looks like we cannot just change the programmatic default (thanks pipeline!)

Note that this would not be 100% safe even for non-Pipeline projects: `readResolve` system only works insofar as you are deserializing an existing `config.xml`. It would not work for people defining projects using (say) `job-dsl`, which while not the common case is not all that exotic either.

jeffret-b commented:
So, what's the necessary level of backwards of compatibility? Can we set the UI to default to not following symlinks and have the constructor default to following? Will that be sufficient?

I'd like to keep this moving forward as best we can to get us into a better position.

mwinter69 committed:
indentation and help text

mwinter69 commented:
> 
> So, what's the necessary level of backwards of compatibility? Can we set the UI to default to not following symlinks and have the constructor default to following? Will that be sufficient?
> 
> I'd like to keep this moving forward as best we can to get us into a better position.

This is how it is at the moment. The UI defaults to false for freestyle and pipeline (snippet generator). As the default is true in the class and readresolve also sets it to true this will ensure full backward compatibility for existing jobs and pipelines I think.

jeffret-b commented:
@Wadeck , you marked this one as needing more reviews and indicated you would review. Would you please add yours when you can?

rsandell commented:
It is a bit unclear to me what the next steps are and if there is any consensus.

jeffret-b commented:
I think it's ready to go. @Wadeck  said it was going to review it and that it needed more reviews so I'm trying to give him the chance. Otherwise, it has three reviews and I don't think it has any outstanding concerns.

Wadeck commented:
I am perfectly fine with one person from our team to have approved it :) (I have change my initial comment to highlight that fact)

daniel-beck commented:
Seems fine. A bit annoying that Pipeline / Job DSL prevent us from really changing the global default to something sane.

Right now,

1. Pipeline and Job DSL will by default _follow_
2. Existing build steps will continue to _follow_ (no behavior change)
3. New build steps will _not follow_ (behavior change)

I'm OK with (3) and it's something we can easily change if people don't like the inconsistency.

jeffret-b commented:
Let's try kicking the CI build.

jeffret-b commented:
CI passes now. Marking ready-for-merge. This can be merged after the waiting period.

jeffret-b commented:
Thanks for the contribution @mwinter69!

daniel-beck commented on bom/pom.xml:
```diff
@@ -59,7 +59,7 @@ THE SOFTWARE.
         <groupId>org.apache.ant</groupId>
         <artifactId>ant</artifactId>
-        <version>1.9.14</version>
+        <version>1.10.7</version>
```
This change breaks `mvn hpi:run` in plugins depending on Jenkins 2.230+ with JDK 8 with the error:

> `Error resolving project artifact: Failure to transfer com.sun:tools:pom:1.8.0` [...]

jeffret-b replied:
We can probably exclude that dependency. It will take more research. Have you filed a Jira ticket or anything @daniel-beck ?

mwinter69 replied:
actually ant doesn't declare this dependency at all. Could this come from the fact that this version of ant was compiled with java11 though with the class versions for java8? If yes we might go back to ant 1.10.5. The reason for using newer ant was that since version 1.10.2 (might also be 1.10.3) ant is able to properly handle symlinks on windows.

jeffret-b replied:
1.10.5 looks like it works from my quick tests. That's a reasonable explanation for where the extra dependency comes from.

daniel-beck replied:
Filed [JENKINS-61937](https://issues.jenkins-ci.org/browse/JENKINS-61937) just now for this.

Weirdly it looked like the dependency exists when I `dependency:tree` the plugin with the Jenkins dependency.
